### DISCUSSION TITLE
[BUG] NuGet: Support build metadata properly

### DISCUSSION THREAD

**Discussion by Happypig375** [community/community#26074] [Actions]
<p>The repo for the following package is <a href="https://github.com/verybadcat/CSharpMath" rel="nofollow noopener">https://github.com/verybadcat/CSharpMath</a>.</p>
<p>After creating a <code>NuGet.Config</code> file as specified (feed is <code>https://nuget.pkg.github.com/verybadcat/index.json</code>), the command <code>dotnet add GitHubReport package CSharpMath.SkiaSharp --version 0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083\+2020.06.10-07.22.37</code> fails:</p>
<pre><code class="lang-auto">  Determining projects to restore...
  Writing C:\Users\hadri\AppData\Local\Temp\tmpB503.tmp
info : Adding PackageReference for package 'CSharpMath.SkiaSharp' into project 'C:\Users\hadri\source\repos\GitHubReport_Nightly\GitHubReport\GitHubReport.csproj'.
info : Restoring packages for C:\Users\hadri\source\repos\GitHubReport_Nightly\GitHubReport\GitHubReport.csproj...
info :   CACHE https://api.nuget.org/v3-flatcontainer/csharpmath.skiasharp/index.json
info :   CACHE https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/index.json
info :   GET https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083/csharpmath.skiasharp.0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083.nupkg
info :   NotFound https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083/csharpmath.skiasharp.0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083.nupkg 1315ms
info :   GET https://api.nuget.org/v3-flatcontainer/csharpmath.skiasharp/index.json
info :   GET https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/index.json
info :   OK https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/index.json 592ms
info :   CACHE https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/index.json
error: The feed 'CSharpMathNightly [https://nuget.pkg.github.com/verybadcat/index.json]' lists package 'CSharpMath.SkiaSharp.0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083' but multiple attempts to download the nupkg have failed. The feed is either invalid or required packages were removed while the current operation was in progress. Verify the package exists on the feed and try again.
error:   Unable to find package 'CSharpMath.SkiaSharp.0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083'.
</code></pre>
<p>The important line is</p>
<pre><code class="lang-auto">info :   NotFound https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083/csharpmath.skiasharp.0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083.nupkg 1315ms
</code></pre>
<p>Opening the URL</p>
<pre><code class="lang-auto">https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083/csharpmath.skiasharp.0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083.nupkg
</code></pre>
<p>, this line is outputted:</p>
<pre><code class="lang-auto">{"error":"Version \"0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083\" does not exist on nuget package \"csharpmath.skiasharp\" under owner \"verybadcat\""}
</code></pre>
<p>This is because</p>
<pre><code class="lang-auto">https://nuget.pkg.github.com/verybadcat/download/csharpmath.skiasharp/0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083+2020.06.10-07.23.29/csharpmath.skiasharp.0.4.2-ci-9db8a6dec29202804764fab9d6f7f19e43c3c083+2020.06.10-07.23.29.nupkg
</code></pre>
<p>is the actual URL.</p>
<p>Contrast this with e.g. Xamarin.Forms on <a href="http://NuGet.org" rel="nofollow noopener">NuGet.org</a>, version</p>
<pre><code class="lang-auto">4.7.0.848-pre3+sha.9bae2b34e-azdo.3764102
</code></pre>
<p>, the download URL is</p>
<pre><code class="lang-auto">https://api.nuget.org/v3-flatcontainer/xamarin.forms/4.7.0.848-pre3/xamarin.forms.4.7.0.848-pre3.nupkg
</code></pre>
<p>, which has NO build metadata in the name.</p>
<p>Therefore, the NuGet GitHub Package registry appending the build metadata to the download URL is incompatible with existing .NET tooling and should be fixed. Thanks.</p>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/happypig375">@happypig375</a>,</p>
<p>I didn’t know NuGet supported SemVer2 and metadata!</p>
<p>I haven’t been able to find much documentation about it. Here are a couple of blog entries and the docs was able to find:<br>
</p><aside class="onebox whitelistedgeneric">
  <header class="source">
      <a href="https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#semantic-versioning-200" target="_blank" rel="nofollow noopener">docs.microsoft.com</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/11719160/181095090-a8aee547-4cb4-4110-aea4-26ecd262286e.png" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#semantic-versioning-200" target="_blank" rel="nofollow noopener">NuGet Package Version Reference</a></h3>

<p>Exact details on specifying version numbers and ranges for other packages upon which a NuGet package depends, and how dependencies are installed.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<br>
<aside class="onebox whitelistedgeneric">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/11719160/181095093-6a434e25-7ac0-4b2d-929c-96b3fd8867a9.png" class="site-icon" width="150" height="150">
      <a href="https://devblogs.microsoft.com/nuget/whats-nu-in-nuget-with-visual-studio-2017-version-15-3/" target="_blank" rel="nofollow noopener" title="04:17PM - 15 August 2017">The NuGet Blog – 15 Aug 17</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/325;"><img src="https://user-images.githubusercontent.com/11719160/181095094-fc589549-9831-4312-a0f8-4e44eba1ba13.png" class="thumbnail" width="690" height="325" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/e/e080b3cf7b6166acb538e71f059efea623d4dd35_2_690x325.png, https://aws1.discourse-cdn.com/github/optimized/2X/e/e080b3cf7b6166acb538e71f059efea623d4dd35_2_1035x487.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/e/e080b3cf7b6166acb538e71f059efea623d4dd35.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/e080b3cf7b6166acb538e71f059efea623d4dd35_2_10x10.png"></div>

<h3><a href="https://devblogs.microsoft.com/nuget/whats-nu-in-nuget-with-visual-studio-2017-version-15-3/" target="_blank" rel="nofollow noopener">What's Nu in NuGet with Visual Studio 2017 version 15.3? | The NuGet Blog</a></h3>

<p>We are happy to announce an update to the NuGet client that comes bundled with Visual Studio 2017 version 15.3 RTW and .NET Core 2.0 SDK. This release introduces support for new scenarios such as .NET Core 2.0/.NET Standard 2.0, some new features,</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<br>
<a href="https://articles.assembla.com/en/articles/2729137-semver2-support-for-nuget-endpoints-on-myget" class="onebox" target="_blank" rel="nofollow noopener">https://articles.assembla.com/en/articles/2729137-semver2-support-for-nuget-endpoints-on-myget</a>
<p>I’m interested to know how you’re using metadata and what problem you’re hoping to solve? Do you want to be able to push multiple build of the same version (like Maven’s SNAPSHOT versions) or are you using it to link back to the commit a package version was built from?</p>
<p>Thanks,<br>
Jamie.</p>

**Comment by Happypig375** [comment]
<p>Multiple versions only differentiable by build metadata are considered the same version. <a href="https://semver.org/#spec-item-10" rel="nofollow noopener">This is how SemVer works.</a> I am using metadata to store build dates - the package commit is encoded in the prerelease section.</p>
<p>The issue is that the GitHub Package registry is incompatible with existing .NET tooling. It should follow <a href="http://NuGet.org" rel="nofollow noopener">NuGet.org</a> in package URL generation.</p>

**Comment by jcansdale** [comment, marked as answer]
<p><a class="mention" href="/happypig375">@happypig375</a> am I understanding this correctly.</p>
<ul>
<li>If a NuGet registry supports storing the <code>+metadata</code> as part of a package version (which GitHub Packages apparently does), it should also support returning the package when the metadata itsn’t metadata</li>
<li>Alternatively a NuGet registry could simply ignore metadata when storing and fetching a package version</li>
</ul>
<p>It sounds like the easy fix would be to simply ignore everything after <code>#</code> in a package version. :thinking:</p>

**Comment by Happypig375** [comment]
<aside class="quote group-github-employees" data-username="jcansdale" data-post="4" data-topic="117606">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>it should also support returning the package when the metadata itsn’t metadata</p>
</blockquote>
</aside>
<p>The metadata IS metadata, no matter the content. Just support it properly, thanks.</p>
<aside class="quote group-github-employees" data-username="jcansdale" data-post="4" data-topic="117606">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>Alternatively a NuGet registry could simply ignore metadata when storing and fetching a package version</p>
</blockquote>
</aside>
<p>If <a href="http://NuGet.org" rel="nofollow noopener">NuGet.org</a> does this, then GitHub Packages should too.</p>

**Comment by jcansdale** [comment]
<blockquote>
<p>The metadata IS metadata, no matter the content. Just support it properly, thanks.</p>
</blockquote>
<p>Sorry, I meant when the metadata isn’t <em>specified</em>. :wink:</p>

**Comment by PathogenDavid** [comment]
<p>I’m also running into this issue. You can’t even manually specify the build metadata as a workaround either since NuGet ignores it entirely.</p>
<p>I made a quick repository that demonstrates the issue using GitHub Actions:</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/278957/181095102-f49880b0-120e-4f7c-9f8f-f9952469a73b.png" class="thumbnail onebox-avatar" width="368" height="368">

<h3><a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest" target="_blank" rel="nofollow noopener">PathogenPlayground/GitHubPackagesBuildMetadataTest</a></h3>

<p>Repo that uses GitHub Actions to demonstrate https://github.community/t/bug-nuget-support-build-metadata-properly/117606/3 - PathogenPlayground/GitHubPackagesBuildMetadataTest</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Basically it publishes two packages to GitHub Packages: One with build metadata and one without. Then in separate jobs it attempts to add those packages to a fresh console project. (With and without explicit build metadata just to show that can’t be used as a workaround.)</p>
<p><a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/actions/runs/192060984" rel="nofollow noopener">Here’s an example run showing the failed restores</a>.</p>
<p>There’s also an additional workflow that demonstrates <a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/actions/runs/192058931" rel="nofollow noopener">everything works as expected when NuGet.org is used instead.</a></p>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/pathogendavid">@PathogenDavid</a>,</p>
<p>Thanks for putting this together!</p>
<p>I noticed your note about <code>source-url</code> needing to be all lowercase.</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/blob/91339050bcbdef86ad9f77df235c930e7c8a42e0/.github/workflows/test.yml#L89" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/blob/91339050bcbdef86ad9f77df235c930e7c8a42e0/.github/workflows/test.yml#L89" target="_blank" rel="nofollow noopener">PathogenPlayground/GitHubPackagesBuildMetadataTest/blob/91339050bcbdef86ad9f77df235c930e7c8a42e0/.github/workflows/test.yml#L89</a></h4>
<pre class="onebox"><code class="lang-yml"><ol class="start lines" start="79" style="counter-reset: li-counter 78 ;">
<li>  with:</li>
<li>    python-version: '3.8'</li>
<li>
</li>
<li>- name: Setup .NET</li>
<li>  uses: actions/setup-dotnet@v1</li>
<li>  with:</li>
<li>    dotnet-version: 3.1.x</li>
<li>
</li>
<li># ----------------------------------------------------------------------- Configure NuGet</li>
<li># We use this script instead of the setup-dotnet settings so that this workflow is owner-agnostic and as such can run in forks.</li>
<li class="selected"># (We can't use github.repository_owner with source-url because it must be all lowercase.)</li>
<li>- name: Configure NuGet</li>
<li>  run: python .github/workflows/make-nuget-config.py</li>
<li>  env:</li>
<li>    GITHUB_REPOSITORY_OWNER: ${{ github.repository_owner }}</li>
<li>    GITHUB_TOKEN: ${{ github.token }}</li>
<li>
</li>
<li># ----------------------------------------------------------------------- Create Project</li>
<li>- name: Create Project</li>
<li>  run: dotnet new console</li>
<li>
</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>I’m interested to know why you think this. An upper-case owner seems work for me, see here:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/jcansdale-test/nuget-errors/runs/944450105?check_suite_focus=true#step:5:8" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/11719160/181095104-e75146f8-f87f-44c4-add3-b4c7b07a078f.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/jcansdale-test/nuget-errors/runs/944450105?check_suite_focus=true#step:5:8" target="_blank" rel="nofollow noopener">jcansdale-test/nuget-errors</a></h3>

<p>Contribute to jcansdale-test/nuget-errors development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by jcansdale** [comment]
<p>I’ve sent you a PR. :slightly_smiling_face:<br>
</p><aside class="onebox githubpullrequest">
  <header class="source">
      <a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/pull/1/files" target="_blank" rel="nofollow noopener">github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/pull/1" target="_blank" rel="nofollow noopener">Configure using actions/setup-dotnet</a>
    </h4>

    <div class="branches">
      <code>PathogenPlayground:main</code> ← <code>jcansdale-test:main</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-08-04" data-time="11:13:58" data-timezone="UTC">11:13AM - 04 Aug 20 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/jcansdale-test" target="_blank" rel="nofollow noopener">
          <img alt="jcansdale-test" src="https://user-images.githubusercontent.com/11719160/181095107-4e04149a-776d-48ed-bb55-b6aafb12abc2.png" class="onebox-avatar-inline" width="20" height="20">
          jcansdale-test
        </a>
      </div>

      <div class="lines" title="10 commits changed 1 files with 16 additions and 37 deletions">
        <a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/pull/1/files" target="_blank" rel="nofollow noopener">
          <span class="added">+16</span>
          <span class="removed">-37</span>
        </a>
      </div>
    </div>

  </div>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by PathogenDavid** [comment]
<aside class="quote group-github-employees" data-username="jcansdale" data-post="8" data-topic="117606">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>Thanks for putting this together!</p>
</blockquote>
</aside>
<p>No problem, hope it helps!</p>
<aside class="quote group-github-employees" data-username="jcansdale" data-post="8" data-topic="117606">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>I’m interested to know why you think this.</p>
</blockquote>
</aside>
<p><a href="https://docs.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-dotnet-cli-for-use-with-github-packages#authenticating-to-github-packages" rel="nofollow noopener">This documentation</a>:</p>
<blockquote>
<p>Because upper case letters aren’t supported, you must use lowercase letters for the repository owner even if the GitHub user or organization name contains uppercase letters.</p>
</blockquote>
<p>I do remember trying ignoring that documentation and using <code>github.repository_owner</code> anyway, but it didn’t work at the time. If I had to guess I was unknowingly suffering from the intermittent failures caused by <a href="https://github.com/NuGet/Home/issues/9775" rel="nofollow noopener">https://github.com/NuGet/Home/issues/9775</a> and blamed the failure on the URL so I made that little Python script to avoid the hard-coded URL.</p>
<aside class="quote group-github-employees" data-username="jcansdale" data-post="9" data-topic="117606">
<div class="title">
<div class="quote-controls"></div>
 jcansdale:</div>
<blockquote>
<p>I’ve sent you a PR. :slightly_smiling_face:</p>
</blockquote>
</aside>
<p>Got it merged and looks like it’s working! I also tested it on my real GitHub Actions consumer and <a href="https://github.com/InfectedLibraries/ClangSharp.Pathogen/runs/944622263?check_suite_focus=true" rel="nofollow noopener">it’s working as well</a>.</p>
<aside class="onebox githubcommit">
  <header class="source">
      <a href="https://github.com/InfectedLibraries/ClangSharp.Pathogen/commit/1b24649c1b006d0a94b375665611de30460c5a6a" target="_blank" rel="nofollow noopener">github.com/InfectedLibraries/ClangSharp.Pathogen</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M10.86 7c-.45-1.72-2-3-3.86-3-1.86 0-3.41 1.28-3.86 3H0v2h3.14c.45 1.72 2 3 3.86 3 1.86 0 3.41-1.28 3.86-3H14V7h-3.14zM7 10.2c-1.22 0-2.2-.98-2.2-2.2 0-1.22.98-2.2 2.2-2.2 1.22 0 2.2.98 2.2 2.2 0 1.22-.98 2.2-2.2 2.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/InfectedLibraries/ClangSharp.Pathogen/commit/1b24649c1b006d0a94b375665611de30460c5a6a" target="_blank" rel="nofollow noopener">[CI] Eliminated make-nuget-config.py</a>
    </h4>

    <div class="github-info">
      <div class="date">
        committed <span class="discourse-local-date" data-format="ll" data-date="2020-08-04" data-time="11:19:46" data-timezone="UTC">11:19AM - 04 Aug 20 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/PathogenDavid" target="_blank" rel="nofollow noopener">
          <img alt="PathogenDavid" src="https://user-images.githubusercontent.com/278957/181095108-c1299075-8a53-4991-988f-48ae50b01c02.jpeg" class="onebox-avatar-inline" width="20" height="20">
          PathogenDavid
        </a>
        
      </div>

      <div class="lines" title="changed 2 files with 3 additions and 44 deletions">
        <a href="https://github.com/InfectedLibraries/ClangSharp.Pathogen/commit/1b24649c1b006d0a94b375665611de30460c5a6a" target="_blank" rel="nofollow noopener">
          <span class="added">+3</span>
          <span class="removed">-44</span>
        </a>
      </div>
    </div>

  </div>
</div>


  <div class="github-row">
    <pre class="github-content" style="white-space: normal;">It turns out that uppercase letters are allowed in the owner name for GitHub Packages.
See discussion here: https://github.community/t/bug-nuget-support-build-metadata-properly/117606/8</pre>
  </div>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by jcansdale** [comment]
<aside class="quote no-group" data-username="PathogenDavid" data-post="10" data-topic="117606">
<div class="title">
<div class="quote-controls"></div>
 PathogenDavid:</div>
<blockquote>
<p>Got it merged and looks like it’s working! I also tested it on my real GitHub Actions consumer and <a href="https://github.com/InfectedLibraries/ClangSharp.Pathogen/runs/944622263?check_suite_focus=true" rel="nofollow noopener">it’s working as well </a>.</p>
</blockquote>
</aside>
<p>Thanks for the confirmation! I’ve suggested we update our docs to reflect this. :slightly_smiling_face:</p>

**Comment by jcansdale** [comment]
<p>The <a href="https://docs.github.com/en/packages/using-github-packages-with-your-projects-ecosystem/configuring-dotnet-cli-for-use-with-github-packages#authenticating-to-github-packages" rel="nofollow noopener">docs</a> have now been updated to <em>not</em> insist on all lowercase! :slightly_smiling_face:</p>

**Comment by Happypig375** [comment]
<p><a class="mention" href="/jcansdale">@jcansdale</a> The original bug mentioned in the top post has yet to be fixed?</p>

**Comment by jcansdale** [comment]
<p><a class="mention" href="/happypig375">@happypig375</a>,</p>
<p>It hasn’t I’m afraid. :slightly_frowning_face:</p>
<p>A possible workaround would be to use the following when publishing to GitHub Packages:</p>
<pre><code class="lang-auto">dotnet tool update gpr -g
gpr push &lt;path to *.nupkg&gt; --version &lt;version without metadata&gt;
</code></pre>
<p>That way you could continue to use metadata on <a href="http://nuget.org" rel="nofollow noopener">nuget.org</a>, but a version stripped on metadata would be published to GitHub Packages.</p>

**Comment by PathogenDavid** [comment]
<p>Thanks for getting the docs issues sorted out!</p>
<p>Unfortunately that workaround didn’t seem to work for me: <a href="https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/runs/1004018759?check_suite_focus=true" rel="nofollow noopener">https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest/runs/1004018759?check_suite_focus=true</a></p>
<p>It appears that <code>gpr</code> is properly re-writing the nuspec to change the version number (<a href="https://github.com/jcansdale/gpr/blob/f6cf8de6b25c44a5ad3ad6cf095147d232097222/src/GprTool/Program.cs#L528" rel="nofollow noopener">this line</a> writes out the right version), but when it actually registers the package the build metadata is back:</p>
<pre><code class="lang-nohighlight">[PackageWithBuildMetadata.1.0.2.nupkg]: Repository url: https://github.com/PathogenPlayground/GitHubPackagesBuildMetadataTest. Version: 1.0.2. Size: 3451 bytes. 
...
[PackageWithBuildMetadata.1.0.2.nupkg]: Successfully registered nuget package: PackageWithBuildMetadata (1.0.2+BuildMetadata)
</code></pre>
<p>I’ll try to find some time to dig further later, but I thought I’d at least share what I discovered in my quick test.</p>

**Comment by jcansdale** [comment]
<p><a class="mention" href="/pathogendavid">@PathogenDavid</a>,</p>
<p>Thanks for the heads up! Could you <a href="https://github.com/jcansdale/gpr/issues" rel="nofollow noopener">open</a> an issue about this to remind me?</p>
<p>It looks like <code>NuGetVersion.ToString</code> doesn’t show metadata by default (which is why the version strings appear different). It would be nice if it did automatically strip any metadata!</p>

**Comment by PathogenDavid** [comment]
<p>Ah, that would certainly do it! Filed an issue here:</p>
<aside class="onebox githubissue">
  <header class="source">
      <a href="https://github.com/jcansdale/gpr/issues/92" target="_blank" rel="nofollow noopener">github.com/jcansdale/gpr</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/jcansdale/gpr/issues/92" target="_blank" rel="nofollow noopener">Specifying the version when pushing doesn't work for stripping build metadata</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-08-20" data-time="00:27:31" data-timezone="UTC">12:27AM - 20 Aug 20 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/PathogenDavid" target="_blank" rel="nofollow noopener">
          <img alt="PathogenDavid" src="https://user-images.githubusercontent.com/278957/181095114-39755dc4-7b20-439a-9e2e-8d5f309d0e5a.jpeg" class="onebox-avatar-inline" width="20" height="20">
          PathogenDavid
        </a>
      </div>
    </div>
  </div>
</div>

<div class="github-row">
  <p class="github-content">As per our discussion here: https://github.community/t/bug-nuget-support-build-metadata-properly/117606/15
It seems gpr push "*.nupkg" --version 1.0.0 doesn't work if the package's version is 1.0.0+BuildMetadata because...</p>
</div>

<div class="labels">
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by jcansdale** [comment]
<p>I’ve created a PR with a fix for this. :crossed_fingers:</p>
<p>Please see the “How to test” section:<br>
</p><aside class="onebox githubpullrequest">
  <header class="source">
      <a href="https://github.com/jcansdale/gpr/pull/93" target="_blank" rel="nofollow noopener">github.com/jcansdale/gpr</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/jcansdale/gpr/pull/93" target="_blank" rel="nofollow noopener">Strip any NuGet version metadata</a>
    </h4>

    <div class="branches">
      <code>jcansdale:master</code> ← <code>jcansdale:fixes/92-strip-metadata</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2020-08-20" data-time="10:14:12" data-timezone="UTC">10:14AM - 20 Aug 20 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/jcansdale" target="_blank" rel="nofollow noopener">
          <img alt="jcansdale" src="https://user-images.githubusercontent.com/11719160/181095116-35a84d1a-b968-4ad7-8446-d5f5ce0af3d9.png" class="onebox-avatar-inline" width="20" height="20">
          jcansdale
        </a>
      </div>

      <div class="lines" title="4 commits changed 3 files with 15 additions and 3 deletions">
        <a href="https://github.com/jcansdale/gpr/pull/93/files" target="_blank" rel="nofollow noopener">
          <span class="added">+15</span>
          <span class="removed">-3</span>
        </a>
      </div>
    </div>

  </div>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>It should automatically strip metadata, so you don’t need to bother with the <code>--version</code> option.</p>
<p>Please let me know how you get on!</p>

**Comment by PathogenDavid** [comment]
<p>I replied on the PR, but for the sake of discussion consistency: I can confirm that fixed my issue. Thanks for getting the workaround fixed!</p>
